File: flutter/hydroponics_framework/lib/datamodels/recipes/phase/controlled_phase_model.dart
import 'package:hydroponics_framework/datamodels/recipes/phase/phase_model.dart';
import 'package:hydroponics_framework/datamodels/recipes/phase/phase_type.dart';

import 'package:hydroponics_framework/datamodels/recipes/process_schedule_model.dart';
import 'package:hydroponics_framework/datamodels/recipes/setpoint_model.dart';

class ControlledPhaseModel extends PhaseModel {
  final List<ProcessScheduleModel>? processSchedules;
  final List<SetpointModel>? setPoints;

  const ControlledPhaseModel({
    required super.phaseId,
    required super.name,
    required super.description,
    required super.duration,
    this.processSchedules,
    this.setPoints
  });

  @override
  Map<String, dynamic> toJson() {
    Map<String, dynamic> json = {};
    json["type"] = PhaseType.controlPhase.name;
    json["phaseId"] = phaseId;
    json["name"] = name;
    json["description"] = description;
    json["duration"] = duration;
    if (processSchedules != null) {
      json["processSchedules"] =
          processSchedules!.map((ps) => ps.toJson()).toList();
    }
    if (setPoints != null) {
      json["setPoints"] = setPoints!.map((sp) => sp.toJson()).toList();
    }
    return json;
  }

  static ControlledPhaseModel fromJson(Map<String, dynamic> json) {
    String phaseId = json["phaseId"];
    String name = json["name"];
    String description = json["description"];
    int duration = json["duration"];

    List<ProcessScheduleModel>? processSchedules;
    if (json["processSchedules"] != null) {
      processSchedules =
          List<Map<String, dynamic>>.from(json["processSchedules"])
              .map((p) => ProcessScheduleModel.fromJson(p)).toList();
    }
    List<SetpointModel>? setPoints;
    if (json["setPoints"] != null) {
      setPoints = List<Map<String, dynamic>>.from(json["setPoints"])
          .map((s) => SetpointModel.fromJson(s)).toList();
    }

    return ControlledPhaseModel(
        phaseId: phaseId,
        name: name,
        description: description,
        duration: duration,
        processSchedules: processSchedules,
        setPoints: setPoints
    );
  }

  SetpointModel? setPointForVariable(String variableId) =>
      setPoints?.firstWhere((sp) => sp.variableId == variableId);


  ProcessScheduleModel? scheduleForScheduler(String schedulerId) =>
      processSchedules?.firstWhere((ps) => ps.schedulerId == schedulerId);
}

File: kotlin/hydroponics-host/hydroponics-host/src/main/kotlin/io/greenglass/hydroponics/application/models/recipe/phase/ControlledPhaseModel.kt
package io.greenglass.hydroponics.application.models.recipe.phase

import org.neo4j.graphdb.Direction.*
import org.neo4j.graphdb.Label
import org.neo4j.graphdb.Node
import org.neo4j.graphdb.RelationshipType
import org.neo4j.graphdb.Transaction

import io.greenglass.host.application.neo4j.GraphNode
import io.greenglass.hydroponics.application.models.recipe.ProcessScheduleModel
import io.greenglass.hydroponics.application.models.recipe.SetpointModel
import io.greenglass.hydroponics.application.models.recipe.SetpointModel.Graph.SET_POINT

class ControlledPhaseModel(
    type : PhaseType,
    phaseId : String,
    name : String,
    description : String,
    duration : Int,
    var processSchedules : List<ProcessScheduleModel>? = null,
    var setPoints : List<SetpointModel>? = null,

    ) : GraphNode, PhaseModel(type, phaseId,name,description, duration) {
    constructor(node : Node) : this(
        PhaseType.ControlPhase,
        node.getProperty("phaseId") as String,
        node.getProperty("name") as String,
        node.getProperty("description") as String,
        node.getProperty("duration") as Int,
        node.getRelationships(OUTGOING, ProcessScheduleModel.PROCESS_SCHEDULE).map { r -> ProcessScheduleModel(r.endNode) },
        node.getRelationships(OUTGOING, SET_POINT).map{ r -> SetpointModel(r.endNode) },
    )
    override fun toNode(tx : Transaction) : Node {
        val node = tx.createNode(phaseLabel, controlPhaseLabel)
        node.setProperty("phaseId", this.phaseId)
        node.setProperty("name", this.name)
        node.setProperty("description", this.description)
        node.setProperty("duration", this.duration)
        processSchedules?.forEach { s -> s.toNode(tx) }
        setPoints?.forEach { sp -> sp.toNode(tx) }
        return node
    }

    companion object Graph {
        val controlPhaseLabel = Label.label("ControlPhase")
        val PHASE : RelationshipType = RelationshipType.withName("PHASE")
    }
}
